### PR TITLE
Add documentation for event filters

### PR DESCRIPTION
**Pull request by Deamon5550**
This adds documentation for [SpongeAPI#927 Event Filters](https://github.com/SpongePowered/SpongeAPI/pull/927)

Please feel free to edit this branch for content/formatting/whatever without asking me for permission.

### CODE REVIEW

**Comment by Spongy** [comment]
A preview for this pull request is available at https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/f4efcebe/index.html.

Here are some links to the pages that were modified:

- added https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/f4efcebe/plugin/events/event-filters.html

_Since the preview frequently changes, please link to [this comment](https://github.com/SpongePowered/SpongeDocs/pull/379#issuecomment-160007573), not to the direct url to the preview._

File: source/plugin/event/event-filters.rst [hunk 1]
```diff
@@ -0,0 +1,145 @@
+======================================
+Advansed Event Handling: Event Filters
```
Comment (Deamon5550) [inline source/plugin/event/event-filters.rst]: Should probably change this to just 'Event Filters'

**Comment by saladoc** [comment]
Open questions of lesser priority: 
 - Can event filters be applied to event listeners registered as per https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/e1a2a6ff/plugin/event/listeners.html#dynamically-registering-event-listeners ? If so, how?
 - `@Listener(ignoreCancelled=false)` vs `@Listener @IsCancelled(CancellationState.IGNORE)` - which one should be preferred?

**Comment by Deamon5550** [comment]
> - Can event filters be applied to event listeners registered as per https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/e1a2a6ff/plugin/event/listeners.html#dynamically-registering-event-listeners ? If so, how?

No they cannot, I toyed briefly with how to do it but there isn't really a way for a) them to cleanly add the filters to their handler since they are implementing a method that they can't add parameters to, and b) us to cleanly generate/transform their handlers to do the filtering. So bottom line, if you're implementing EventListener yourself you're on your own for filtering.

 > - `@Listener(ignoreCancelled=false)` vs `@Listener @IsCancelled(CancellationState.IGNORE)` - which one should be preferred?

`@Listener(ignoreCancelled=false)` is going to be removed soon and `@IsCancelled` will be the only place for controlling behavior around cancellation. I'll update this once I make that change.

File: source/plugin/event/index.rst [hunk 2]
```diff
@@ -0,0 +1,39 @@
+===================
+Working with Events
+===================
+
+Sponge provides a system to:
+
+- Listen for events
+- Fire events
+
+Overview
```
Comment (Tzky) [inline source/plugin/event/index.rst]: I'd say we should get rid of the ``Overview`` heading and the list above. Just some introductory sentences should suffice.

**Comment by Tzky** [comment]
Is the concept of our events explained anywhere (``Targets``+``Causes``)? Same for the naming scheme...

**Comment by saladoc** [comment]
The events page we have mostly deals with listeners. Causes I introduced as a stub page, thats #267.

**Comment by Deamon5550** [comment]
I'll write a page for causes sometime today

**Comment by saladoc** [comment]
Looks good so far. Maybe mention that `NamedCause` also holds constants holding common cause names so you don't have to toss around magic values.

**Comment by Deamon5550** [comment]
@saladoc I added a note to the bottom of the named cause section https://github.com/SpongePowered/SpongeDocs/pull/379/files#diff-dba86822fc60edfa471b8efba4f5b7e2R85

File: source/plugin/event/causes.rst [hunk 3]
```diff
@@ -0,0 +1,121 @@
+============
+Event Causes
+============
+
+Events are great for the use case of wanting to execute something when certain actions occur, but
+they have the drawback of providing next to no context as to what has **caused** that event to
+occur. A ``Cause`` allows providing and receiving additional contextual information about the event
+which can be do to modify the behavior of your event handler depending on this context.
```
Comment (Inscrutable) [inline source/plugin/event/causes.rst]: Something is wrong with this sentence, "which can be do to modify". We can clean up some wording later, but I'm really not sure what you meant to say here.

File: source/plugin/event/causes.rst [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+conditions this information is instead provided in the ``Cause`` of the event.
+
+Any event which extends ``CauseTracked`` provides a ``Cause`` object which can be interogated
+for the information pertaining to why the event was fired. The Cause object can be retrived from
+an event by simply calling ``event.getCause()``.
+
+Retrieving objects from a Cause
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Structureally a ``Cause`` object contains a sequential list of objects. There are several methods of
```
Comment (Inscrutable) [inline source/plugin/event/causes.rst]: "Structurally," (typo and add a comma). In general add a lot more commas.

File: source/plugin/event/causes.rst [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    Some common identifying names for named causes are present as static fields in the
+    ``NamedCause`` class. Identifiers which are specific to certain events can often be found
+    as static fields on the event class, for example ``DamageEntityEvent.SOURCE``.
+
+Creating custom Causes
+~~~~~~~~~~~~~~~~~~~~~~
+
+Creating a cause to use when firing an event is extremely easy. The hardest part is deciding
+what information to include in the cause. If you're firing an event from your plugin which is
+usually triggered through other means perhaps you want to include your plguin container so
```
Comment (Inscrutable) [inline source/plugin/event/causes.rst]: "plugin"

File: source/plugin/event/causes.rst [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+other plugins know that the event comes from your plugin. Or if you are firing the event on
+behalf of a player due to some action its usually a good idea to include that player in
```
Comment (Inscrutable) [inline source/plugin/event/causes.rst]: the catastrophe of the missing apostrophe. "it's" (it is).

File: source/plugin/event/causes.rst (hunk 3 above)
Comment (Deamon5550) [inline source/plugin/event/causes.rst]: I think I meant to say `...which can be used to modify...`

File: source/plugin/event/filters.rst [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+.. code-block:: java
+
+    @Listener
+    public void onInteract(InteractBlockEvent.Secondary event, @First @Has(FlyingData.class) Entity entity) {
+        // do something
+    }
+
+.. note::
+    Both ``@Has`` and ``@Supports`` have an optional parameter ``inverse()`` which can be set to cause validation
```
Comment (boformer) [inline source/plugin/event/filters.rst]: maybe remove the brackets?

**Comment by boformer** [comment]
Looks good.

**Pull request merged**